Compatibility with BSD date

gamma opened the pull request:
The macOS shell has a different date format and does not support the uppercase shortcut.

gamma committed:
Compatibility with macOS shell

The macOS shell has a different date format and does not support the uppercase shortcut.

oh2fih commented:
Thanks! Looks good. I'll just test this still works as intended on Linux.

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -82,7 +86,7 @@ else
       | grep -Eo 'expires:\ [0-9\-]+' \
       | awk '{ print $2}' \
-      | date -Iseconds -u -f - \
+      | ( [[ IS_MAC ]] && date -Iseconds -u || date -Iseconds -u -f - ) \
```
The `[[ IS_MAC ]]` is always `true`.

```suggestion
      | ( [[ "$IS_MAC" = "true" ]] && date -Iseconds -u || date -Iseconds -u -f - ) \
```

See:

```
$ IS_MAC=true ; echo "foo" | ([[ $IS_MAC ]] && sed 's/foo/true/' || sed 's/foo/false/')
true
$ IS_MAC=false ; echo "foo" | ([[ $IS_MAC ]] && sed 's/foo/true/' || sed 's/foo/false/')
true

$ IS_MAC=true ; echo "foo" | ([[ "$IS_MAC" = "true" ]] && sed 's/foo/true/' || sed 's/foo/false/')
true
$ IS_MAC=false ; echo "foo" | ([[ "$IS_MAC" = "true" ]] && sed 's/foo/true/' || sed 's/foo/false/')
false
```

I think the same mistake elsewhere causes the `Expires:` field being empty on Linux.

oh2fih replied:
Updated suggestion per the other comment with `1` & `0`:
```suggestion
      | ( (( IS_MAC )) && date -Iseconds -u || date -Iseconds -u -f - ) \
```

oh2fih replied:
On the other hand we should consider [SC2015](https://www.shellcheck.net/wiki/SC2015):

> Note that `A && B || C` is not if-then-else. C may run when A is true.

This might require a full if-then-else:

```bash
  if (( IS_MAC )); then
    KEY_EXPIRES=$(
      echo "$KEY_INFO" \
        | grep "sec" \
        | grep -Eo 'expires:\ [0-9\-]+' \
        | awk '{ print $2}' \
        | date -Iseconds -u \
        | sed -e 's/+00:00$/Z/'
      )
  else
    KEY_EXPIRES=$(
      echo "$KEY_INFO" \
        | grep "sec" \
        | grep -Eo 'expires:\ [0-9\-]+' \
        | awk '{ print $2}' \
        | date -Iseconds -u -f - \
        | sed -e 's/+00:00$/Z/'
      )
  fi
```

oh2fih replied:
As I have fixed the ShellCheck pipeline in `master` it now correctly fails.

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -33,6 +33,10 @@ if [ "$#" -lt 1 ]; then
 
+# Prepare environment
+
+[[ "$(uname)" == "Darwin" ]] && IS_MAC=true || IS_MAC=false
```
Bash doesn't really have booleans. If you used `1` &  `0` instead, you could use arithmetic expressions.

```suggestion
[[ "$(uname)" == "Darwin" ]] && IS_MAC=1 || IS_MAC=0
```

Then:

```
$ IS_MAC=1 ; (( IS_MAC )) && echo "true" || echo "false"
true
$ IS_MAC=0 ; (( IS_MAC )) && echo "true" || echo "false"
false

oh2fih replied:
On the other hand we should consider [SC2015](https://www.shellcheck.net/wiki/SC2015):

> Note that `A && B || C` is not if-then-else. C may run when A is true.

This might require a full if-then-else.

oh2fih replied:
If `uname` on macOS has `-s`, this could explicitly be `$(uname -s)` 🤔, but that is a minor improvement.

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -96,8 +100,8 @@ else
   GREPABLE_KEY=${KEY//0x/}
   FP=$(echo "$KEY_INFO" | grep -i "$GREPABLE_KEY" | sed -e 's/[^A-F0-9]//g')
   KEY="0x${FP}"
-  if ! [[ "${KEY}" = "0x${GREPABLE_KEY^^}" ]]; then
```
Do you know why this didn't work on macOS?

gamma replied:
The error is: `securitytxt-signer.sh: line 104: \033[0;33mEXPANDED 0x${GREPABLE_KEY^^} TO ${KEY}\033[0;0m: bad substitution`

Maybe a problem with my bash:

```
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin23)
Copyright (C) 2007 Free Software Foundation, Inc.
```

oh2fih replied:
That Bash is ancient.
```
GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2020 Free Software Foundation, Inc.
```

oh2fih replied:
This script already uses `sed` & `awk` and this uppercase conversion can be done with both. Despite they are all basic UNIX CLI tools I generally prefer keeping the set of commands minimal.

oh2fih replied:
Using `sed` would come back to [SC2001](https://www.shellcheck.net/wiki/SC2001) which was already solved with [Shell Parameter Expansion](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html). Therefore `awk` might be the tool to use here, but please test it with macOS, too.

oh2fih replied:
This `${variable^^}` shell expansion was introduced in Bash 4.0. From the changelog:

```
hh. There are new case-modifying word expansions: uppercase (^[^]) and
    lowercase (,[,]).  They can work on either the first character or
    array element, or globally.  They accept an optional shell pattern
    that determines which characters to modify.  There is an optionally-
    configured feature to include capitalization operators.
 ```

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -115,13 +119,22 @@ fi
 # If the key expires before the DAYS_MAX, use the key expiration date instead.
 
-EXPIRES=$(date -Iseconds -u -d "${DAYS_MAX} days" | sed -e 's/+00:00$/Z/')
+if [[ IS_MAC ]]; then
```
```suggestion
if (( IS_MAC )); then
```

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -115,13 +119,22 @@ fi
 else
-  EXPIRES_COMPARABLE=$(date -d "$EXPIRES" +%s)
-  KEY_EXPIRES_COMPARABLE=$(date -d "$KEY_EXPIRES" +%s)
+  if [[ IS_MAC ]]; then
```
```suggestion
  if (( IS_MAC )); then
```

oh2fih committed:
Apply suggestions from code review

use arithmetic expressions in macOS recognition

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -96,8 +100,8 @@ else
   KEY="0x${FP}"
-  if ! [[ "${KEY}" = "0x${GREPABLE_KEY^^}" ]]; then
-    echo -e "\033[0;33mEXPANDED 0x${GREPABLE_KEY^^} TO ${KEY}\033[0;0m"
+  if ! [[ "${KEY}" = "0x$(echo "${GREPABLE_KEY}" | tr '[:lower:]' '[:upper:]')" ]]; then
```
```suggestion
  if ! [[ "${KEY}" = "0x$(echo "${GREPABLE_KEY}" | awk '{print toupper($0)}')" ]]; then
```

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -96,8 +100,8 @@ else
-  if ! [[ "${KEY}" = "0x${GREPABLE_KEY^^}" ]]; then
-    echo -e "\033[0;33mEXPANDED 0x${GREPABLE_KEY^^} TO ${KEY}\033[0;0m"
+  if ! [[ "${KEY}" = "0x$(echo "${GREPABLE_KEY}" | tr '[:lower:]' '[:upper:]')" ]]; then
+    echo -e "\033[0;33mEXPANDED 0x$(echo "${GREPABLE_KEY}" | tr '[:lower:]' '[:upper:]') TO ${KEY}\033[0;0m"
```
```suggestion
    echo -e "\033[0;33mEXPANDED 0x$(echo "${GREPABLE_KEY}" | awk '{print toupper($0)}') TO ${KEY}\033[0;0m"
```

oh2fih committed:
Replace tr with awk

Awk is already used elsewhere in the script

oh2fih committed:
Fix SC2015

Use full if-then-else

oh2fih commented:
I have fixed the issues I found. Please test whether this is still compatible with macOS.

oh2fih commented on securitytxt-signer.sh:
```diff
@@ -77,14 +81,25 @@ if ! [[ "$KEY" =~ ^0x[a-fA-F0-9]{8,40}$ ]]; then
+        | grep "sec" \
+        | grep -Eo 'expires:\ [0-9\-]+' \
+        | awk '{ print $2}' \
+        | date -Iseconds -u \
```
Does this really take the time from the pipe and use that? Based on https://ss64.com/mac/date.html it would only print the **current** date. We don't want it to expire **now** but either after `$DAYS_MAX` or when the key expires; whichever comes sooner.

oh2fih replied:
"Providing arguments will format the date and time in a user-defined way"; will work, but I wouldn't detect MacOS but BSD version of Date.

oh2fih committed:
Require Bash >= 4.0

The Bash on MacOS X is ancient 3.2;
trying to support it will limit development of this script.

oh2fih committed:
Detect non-GNU date instead of MacOS

oh2fih committed:
Revert back to Parameter Expansion

Now that Bash < 4.0 is not supported, we can revert this.

oh2fih committed:
Update README.md

Add compatibility information.

oh2fih commented:
This PR isn't anymore about Mac OS X shell compatibility as it originally was. Because Apple has [chosen `zsh` as the default shell for Mac](https://support.apple.com/en-us/102360), trying to make Bash scripts compatible with it would require compatibility with ancient version of Bash. Therefore, a better approach is to require Bash >= 4.0; it can be [updated on a Mac](https://www.unindented.org/blog/change-shell-to-latest-bash-on-macos/#installing-the-latest-version-of-bash) from [Homebrew](https://brew.sh/).

However, **your contribution is still accepted** with some modifications **as a support for the BSD `date`**. 🎉
